src/action/filesystems.rs

```diff
@@ -0,0 +1,251 @@
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the filesystems action.
+//!
+//! The filesystems action lists all mounted filesystems on the client,
+//! collecting device name, mount point, filesystem type and its options.
+//! Current implementation works only in Linux systems.
+
+use rrg_proto::{Filesystem, KeyValue, AttributedDict, DataBlob};
+use crate::session::{self, Session};
+
+use log::error;
+use std::fmt::{Display, Formatter};
+
+/// Enum of possible errors, which can occur during collecting filesystems data.
+#[derive(Debug)]
+enum Error {
+    /// Missed mtab-like file error.
+    MissingFile(std::io::Error),
+    /// Parsing mtab-like file error.
+    MountInfoParse(std::io::Error),
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            MissingFile(ref error) => Some(error),
+            MountInfoParse(ref error) => Some(error),
+        }
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use Error::*;
+
+        match *self {
+            MissingFile(ref error) => {
+                write!(fmt, "missed file error: {}", error)
+            },
+            MountInfoParse(ref error) => {
+                write!(fmt, "failed to obtain mount information: {}", error)
+            },
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+/// A response type for the filesystems action.
+pub struct Response {
+    /// Information about the filesystem.
+    mount_info: proc_mounts::MountInfo,
+}
+
+/// Handles requests for the filesystems action.
+///
+/// Initially searches in `/proc/mounts`. If it's missed, falls back to
+/// `/etc/mtab`.
+pub fn handle<S: Session>(session: &mut S, _: ()) -> session::Result<()> {
+    use proc_mounts::MountIter;
+
+    // Check whether `/proc/mounts` exists.
+    let mount_iter = match MountIter::new() {
+        Ok(mount_iter) => mount_iter,
+        Err(_) => {
+            // `/proc/mounts` doesn't exist. Try to fall back to `/etc/mtab`.
+            MountIter::new_from_file("/etc/mtab").map_err(Error::MissingFile)?
+        },
+    };
+
+    for mount_info in mount_iter {
+        let mount_info = mount_info.map_err(Error::MountInfoParse)?;
+        session.reply(Response {
+           mount_info,
+        })?;
+    }
+
+    Ok(())
+}
+
+/// Converts filesystem mount option in `String` representation to
+/// GRR's `KeyValue` protobuf struct representation.
+fn option_to_key_value(option: String) -> KeyValue {
+    match &option.split('=').collect::<Vec<&str>>()[..] {
+        &[key] => {
+            KeyValue {
+                k: Some(DataBlob {
+                    string: Some(String::from(key)),
```

Comment:
Could you add a `TODO` comment about fixing ergonomics of working with `DataBlob`s? Just to have a reminder for myself in code. It looks overly verbose for something as simple as creating key-value pair, I should improve it one day.

Reply:
Sure, I added.
This verbosity is due to `DataBlob`'s universality of supporting many types.